### PR TITLE
Account for grid shift in HPMC external fields

### LINKED ISSUE #1562: Unexpected behavior with CPPExternalPotential
### Description

I am running an HPMC simulation with a custom external potential using `hpmc.external.user` and am getting unexpected results. The potential should bias particle positions to a specific z-value, but that is not happening. The particles do not appear to be influenced by the external potential.

@joaander found the problem: the functions in the `ExternalFieldJIT` C++ class do not account for the grid shift when computing energies. I will implement a bugfix very soon.

### Script

```python
import hoomd
```

### Input files

_No response_

### Output

```shell
No output.
```

### Expected output

_No response_

### Platform

CPU, Linux

### Installation method

Compiled from source

### HOOMD-blue version

3.11.0

### Python version

3.9.15

### PR DESCRIPTION
**Pull request by tcmoore3**
## Description

Account for grid shift in member functions of C++ class `ExternalFieldJIT`.

## Motivation and context

Fixes a bug.

Resolves #1562 

## How has this been tested?

I updated the `test_gravity` unit test to be more robust. I have also observed the correct behavior in the simulation mentioned in #1562. 

## Change log

```
- Fix bug in hpmc.external.user.CPPExternalPotential
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/hpmc/pytest/test_external_user.py
Comment (joaander) [inline hoomd/hpmc/pytest/test_external_user.py:1]: We discussed improvements to the test offline.

File: hoomd/hpmc/ExternalFieldJIT.h [hunk 1]
```diff
@@ -161,7 +163,12 @@ template<class Shape> class ExternalFieldJIT : public hpmc::ExternalFieldMono<Sh
             // read in the current position and orientation
             Scalar4 postype_i = h_postype.data[i];
             unsigned int typ_i = __scalar_as_int(postype_i.w);
-            vec3<Scalar> pos_i = vec3<Scalar>(postype_i);
+            int3 image = make_int3(0, 0, 0);
+            vec3<Scalar> pos_i = vec3<Scalar>(postype_i) - vec3<Scalar>(this->m_pdata->getOrigin());
+            box_new.wrap(pos_i, image);
+            image = make_int3(0, 0, 0);
+            vec3<Scalar> old_pos_i = vec3<Scalar>(*(position_old + i));
```
Comment (joaander) [inline hoomd/hpmc/ExternalFieldJIT.h]: * `old_pos_i` must be shifted by the old origin, which the caller must provide.
* `IntegratorHPMC::attemptBoxResize` should scale the origin.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks for the fix. I have suggestions for improvement.

File: hoomd/hpmc/IntegratorHPMC.cc [hunk 2]
```diff
@@ -153,6 +153,12 @@ bool IntegratorHPMC::attemptBoxResize(uint64_t timestep, const BoxDim& new_box)
 
     m_pdata->setGlobalBox(new_box);
 
+    // scale the origin
+    Scalar3 old_origin = m_pdata->getOrigin();
```
Comment (joaander) [inline hoomd/hpmc/IntegratorHPMC.cc:157]: Excellent. Please make similar changes to `BoxResizeUpdater.cc`.

File: hoomd/hpmc/UpdaterBoxMC.cc [hunk 3]
```diff
@@ -305,6 +311,9 @@ inline bool UpdaterBoxMC::box_resize_trial(Scalar Lx,
 
         m_pdata->setGlobalBox(curBox);
 
+        // reset origin
+        m_pdata->translateOrigin(-origin_shift);
```
Comment (joaander) [inline hoomd/hpmc/UpdaterBoxMC.cc:315]: Excellent. Please make similar changes to `UpdaterQuickCompress.cc`.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! This looks to be exactly what we need to fix external fields in HPMC with box moves.

I would like to see a test for this. First let's look at the trajectory of a simulation run (not a unit test) and see if we can identify a way to write a unit test. You could use your existing harmonic test, though I think it the results might be more clear if we instead place disks outside of a cylindrical wall. 

Before this fix: when run in MPI, the origin will shift inconsistently with the particles, so at some point particles will either end up stuck inside the cylinder and/or the cylinder may appear to diffuse through the simulation box. After this fix, the trajectory should look normal with particles moving around a fixed cylinder.

**Comment by tcmoore3** [comment]
> Thanks! This looks to be exactly what we need to fix external fields in HPMC with box moves.
> 
> I would like to see a test for this. First let's look at the trajectory of a simulation run (not a unit test) and see if we can identify a way to write a unit test. You could use your existing harmonic test, though I think it the results might be more clear if we instead place disks outside of a cylindrical wall.
> 
> Before this fix: when run in MPI, the origin will shift inconsistently with the particles, so at some point particles will either end up stuck inside the cylinder and/or the cylinder may appear to diffuse through the simulation box. After this fix, the trajectory should look normal with particles moving around a fixed cylinder.

@joaander and I worked out a test that highlights the unexpected behavior, which I will summarize here for future reference. The test system is a system of hard spheres initially at a low packing fraction with an HPMC integrator with move size 0 and a `hpmc.update.BoxMC` updater set to `betaP = 10`. As the simulation runs, the box updater causes the system density to increase. Before the fix (specifically, using the v3.10 release), the grid shift is apparent if you track the trajectories of individual particles as the system compresses; the trajectories appear as wavy lines. In contrast, the trajectories are straight lines with this fix. The attached figure illustrates these differences. Note that in this example, only the "dynamics" are affected and the thermodynamics are still correct. However, in cases where a potential energy function depends on spatial coordinates, as discussed in #1562, the thermodynamics of the system are also incorrect.
image

This fix may also resolve the behavior described in #568.

**Review by joaander — APPROVED** [review]
Thanks!

**Pull request merged**